feature/text_based.py

```diff
@@ -0,0 +1,694 @@
+            print("=" * 40)
+
+        return selected
+
+    def _select_random(self, trials: int = 10) -> List[int]:
+        # Set the seed
+        random.seed(self.seed)
+        best_selected = []
+        best_covered = 0
+
+        if self.selection_size is None:
+            unicost, selected_size = self._get_selected_size()
+        else:
+            selected_size = self.selection_size
+
+        for t in range(trials):
+            # Select a sample of selection_size without repetition
+            selected = []
+            while len(selected) < selected_size and self._num_cols is not None:
+                i = random.randint(0, self._num_cols - 1)
+                while i in selected:
+                    i = random.randint(0, self._num_cols -1)
+                selected.append(i)
+
+            # Count covered categories
+            num_row_covered = self._get_num_row_covered(selected)
+            if num_row_covered > best_covered:
+                best_covered = num_row_covered
+                best_selected = selected
+
+        # Calculate coverage metrics
+        num_row_covered = self._get_num_row_covered(best_selected)
+        coverage = num_row_covered / self._num_rows
+
+        if self.verbose:
+            print("\nRANDOM SELECTION:", selected_size, "columns to cover rows ", self._num_rows)
+            print("=" * 40)
+            print("SIZE:", len(best_selected),
+                  "reduction: {:.2f}".format((self._num_cols - len(best_selected)) / self._num_cols))
+            print("SELECTED:", best_selected)
+            print("NUM (AVG) ROWS COVERED:",
+                  num_row_covered, "coverage: {:.2f}".format(coverage))
+            print("STATUS: RANDOM")
+            print("Cost Metric:", self.cost_metric)
+            print("=" * 40)
+        return best_selected
+
+    def _solve_set_cover(self, data: Data) -> List:
+
+        # Create Model object
+        model = Model("Set Cover Model")
+
+        # Variables
+        x = [model.add_var(var_type=BINARY) for _ in data.X]
+
+        # Constraint: every row should be covered
+        for row in data.rows:
+            model.add_constr(xsum(data.matrix[row, i] * x[i] for i in data.X) >= 1)
+
+        # Objective: minimize
+        model.objective = minimize(xsum(data.cost[i] * x[i] for i in data.X))
+
+        # Solve (optimize using Gurobi)
+        model.verbose = 0
+        model.optimize()
+        check_true(model.status == OptimizationStatus.OPTIMAL, "Max Cover Error: optimal solution not found.")
+
+        # Solution
+        selected = [i for i in data.X if float(x[i].x) >= 0.99]
+
+        if self.verbose:
+            print("=" * 40)
+            print("SET COVER OBJECTIVE:", model.objective_value)
+            print("SELECTED:", selected)
+            print("STATUS:", model.status)
+            print("=" * 40)
+
+        # Return
+        return selected
+
+    def _solve_set_cover_cost_option(self, cost: np.ndarray) -> List:
```

Comment:
I am not a big fan of this function.. The reason we have the DATA class is not pass back and forth individual elements. I would rather replace Data.cost outside and just call solve_set_cover(data) and not have this function 

(Notice the "implicit" use of "matrix" inside which is very obscure from outside to understand what is this solving)

Reply:
The function has been removed. The reason of adding this function was in _select_multi_level_optimization(), we need to do Data.cost and call solve_set_cover(data) multiple times for if/else statements.